quaint/src/connector/postgres/native/cache.rs

```diff
@@ -0,0 +1,388 @@
+    cache: Mutex<LruCache<QueryKey, V>>,
+    state: RandomState,
+}
+
+impl<V> InnerLruCache<V> {
+    fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cache: Mutex::new(LruCache::new(capacity)),
+            state: RandomState::new(),
+        }
+    }
+
+    async fn get(&self, sql: &str, types: &[Type]) -> Option<V>
+    where
+        V: Clone,
+    {
+        let mut cache = self.cache.lock().await;
+        let capacity = cache.capacity();
+        let stored = cache.len();
+
+        let key = QueryKey::new(&self.state, sql, types);
+        // we call `get_mut` because LRU requires mutable access for lookups
+        match cache.get_mut(&key) {
+            Some(value) => {
+                tracing::trace!(
+                    message = "CACHE HIT!",
+                    query = sql,
+                    capacity = capacity,
+                    stored = stored,
+                );
+                Some(value.clone())
+            }
+            None => {
+                tracing::trace!(
+                    message = "CACHE MISS!",
+                    query = sql,
+                    capacity = capacity,
+                    stored = stored,
+                );
+                None
+            }
+        }
+    }
+
+    pub async fn insert(&self, sql: &str, types: &[Type], value: V) {
+        let key = QueryKey::new(&self.state, sql, types);
+        self.cache.lock().await.insert(key, value);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use std::future::Future;
+
+    pub(crate) use crate::connector::postgres::url::PostgresNativeUrl;
+    use crate::{
+        connector::{MakeTlsConnectorManager, PostgresFlavour},
+        tests::test_api::postgres::CONN_STR,
+    };
+    use url::Url;
+
+    #[tokio::test]
+    async fn noop_cache_returns_new_queries_every_time() {
+        run_with_client(|client| async move {
+            let cache = NoOpCache;
+            let sql = "SELECT $1";
+            let types = [Type::INT4];
+
+            let stmt1 = cache.get_query(&client, sql, &types).await.unwrap();
+            let stmt2 = cache.get_query(&client, sql, &types).await.unwrap();
+            assert_ne!(stmt1.name(), stmt2.name());
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn noop_cache_returns_new_statements_every_time() {
+        run_with_client(|client| async move {
+            let cache = NoOpCache;
+            let sql = "SELECT $1";
+            let types = [Type::INT4];
+
+            let stmt1 = cache.get_statement(&client, sql, &types).await.unwrap();
+            let stmt2 = cache.get_statement(&client, sql, &types).await.unwrap();
+            assert_ne!(stmt1.name(), stmt2.name());
+        })
+        .await;
+    }
+
+    #[tokio::test]
+    async fn prepared_statement_lru_cache_reuses_queries_within_capacity() {
+        run_with_client(|client| async move {
+            let cache = PreparedStatementLruCache::with_capacity(1);
```

Comment:
Let's expand this test case with a higher capacity, e.g., 3.

Reply:
done